metagraph/src/annotation/row_diff_builder.cpp

```diff
@@ -378,34 +391,49 @@ void build_pred_succ(const std::string &graph_fname,
-            if (!dummy[boss_idx]) {
-                BOSS::edge_index next = boss.fwd(boss_idx);
-                assert(next);
-                if (!dummy[next]) {
-                    while (rd_succ.size() && !rd_succ[next]) {
-                        next--;
-                        assert(!boss.get_last(next));
+            if (succinct) { // Legacy code for DBGSuccinct
+                const BOSS &boss = succinct->get_boss();
+                BOSS::edge_index boss_idx = succinct->kmer_to_boss_index(i);
+                if (!(*dummy)[boss_idx]) {
+                    BOSS::edge_index next = boss.fwd(boss_idx);
+                    assert(next);
+                    if (!(*dummy)[next]) {
+                        while (rd_succ.size() && !rd_succ[next]) {
+                            next--;
+                            assert(!boss.get_last(next));
+                        }
+                        succ_buf.push_back(to_row(succinct->boss_to_kmer_index(next)));
+                        succ_boundary_buf.push_back(0);
                     }
-                    succ_buf.push_back(to_row(graph.boss_to_kmer_index(next)));
-                    succ_boundary_buf.push_back(0);
-                }
-                // compute predecessors only for row-diff successors
-                if (rd_succ.size() ? rd_succ[boss_idx] : boss.get_last(boss_idx)) {
-                    BOSS::TAlphabet d = boss.get_node_last_value(boss_idx);
-                    BOSS::edge_index back_idx = boss.bwd(boss_idx);
-                    boss.call_incoming_to_target(back_idx, d,
-                        [&](BOSS::edge_index pred) {
-                            // dummy predecessors are ignored
-                            if (!dummy[pred]) {
-                                uint64_t node_index = graph.boss_to_kmer_index(pred);
-                                pred_buf.push_back(to_row(node_index));
-                                pred_boundary_buf.push_back(0);
+                    // compute predecessors only for row-diff successors
+                    if (rd_succ.size() ? rd_succ[boss_idx] : boss.get_last(boss_idx)) {
+                        BOSS::TAlphabet d = boss.get_node_last_value(boss_idx);
+                        BOSS::edge_index back_idx = boss.bwd(boss_idx);
+                        boss.call_incoming_to_target(back_idx, d,
+                            [&](BOSS::edge_index pred) {
+                                // dummy predecessors are ignored
+                                if (!(*dummy)[pred]) {
+                                    uint64_t node_index = succinct->boss_to_kmer_index(pred);
+                                    pred_buf.push_back(to_row(node_index));
+                                    pred_boundary_buf.push_back(0);
+                                }
                             }
-                        }
-                    );
+                        );
+                    }
+                }
```

Comment:
I need some external opinion on whether this is actually equivalent to the new code below, and whether we want to preserve this legacy code that works with BOSS directly. For row_diff_traverse and row_diff_successor I simply make overloads in DBGSuccinct that call them in BOSS instead of using base class implementations. I assume we might want to factor this piece out into a separate function as well?

Reply:
It's important to keep this legacy code since we don't preserve a `dummy` vector for BOSS tables to save space. We can also factor this out into a helper function if it cleans things up.

Reply:
Actually, upon reading this code again I don't really understand why we want to preserve it.

1. After doing `boss_idx = succinct->kmer_to_boss_index(i)`, isn't the check that `boss_idx` is dummy redundant? I thought dummy nodes are those that are outside of the graph nodes, so it shouldn't be possible to get mapped into one.
2. What follows within the `if` seems to be equivalent to `row_diff_successor` call, so why not just call it instead of re-implementing it?
3. The part of finding predecessors also seems to be equivalent to `adjacent_incoming_nodes`, because `adjacent_incoming_nodes` presumably shouldn't call any dummy nodes, but only actual nodes from the graph? I'm also not sure if this specific implementation is more efficient than `adjacent_incoming_nodes`?

Reply:
`kmer_to_boss_index` is the identity map if a dummy edge mask is not present, so the dummy edge checks are not done. Also, `adjacent_incoming_nodes` and `adjacent_outgoing_nodes` don't do dummy edge checks if a mask is not present.

Reply:
But I agree with (2), we can replace that block with `row_diff_successor`

Reply:
UPD: Regarding 1, it seems that there are nodes such that `boss_idx` is not dummy, but `boss.fwd(boss_idx)` is dummy. This is very confusing to me, why does this ever happen?

Reply:
I'll close this for now, let's move to that refer to the current state of the code.